### ISSUE TITLE
Debugger Feature Request:  Mitigate the bounce factor while scrolling upwards

### ISSUE THREAD

**Issue by TommyGH** [AppleWin/AppleWin#475]
With all the problems I've been experiencing recently while using the AppleWin debugger, I thought I hit upon some more strange behavior.  Upon further reflection, I now believe the debugger is working "as designed", so this issue is a feature request

I have attached a simple debugger script (debug2.txt):

[debug2.txt](https://github.com/AppleWin/AppleWin/files/1236618/debug2.txt)

1. Start AppleWin
2. Boot DOS3.3 disk (not necessary, but it does help in this case)
3. F7
4. RUN debug2.txt
5. At this point you should be positioned with the VDFILE symbol at top of the screen.  Use the "up arrow" key on your keyboard (or use your mouse wheel) to navigate to the symbol VSPARE.

Try not getting mesmerized while watching the display.

Obviously scrolling downwards from VSPARE is boring compared to the reverse trip.

I realize I can alter my "db" "db8" and "dw" definitions to help with this "problem".

So, is there anything that can be done to help prevent the display from jumping up and down so much?  Or to prevent the defined symbols from appearing and then disappearing while scrolling upwards (other than defining each with its own "db" or "dw", etc.)?  And in my current work environment, the SYMUSER commands are not in my debug script, but are defined in a SYM file.

Tommy

**Comment by Michaelangel007** [comment]
> So, is there anything that can be done to help prevent the display from jumping up and down so much? 

Generally, no, this problem is unsolvable.  It was one of the first things I tried to address when I joined the AppleWin team.

Here is why it comes about.

The cursor line is trying to be shown in the middle of the pane at all times in order to show the user the "context".

Sometimes this fails due to the contiguous opcodes forcing it to get "out-of-sync" -- at which time it is shown at the first line.

There will always be edge cases that can force this to fail.

The only _partial_ solution is to always show the cursor line on the 1st row -- but then you lose the preceding context.

Marking this as "Won't Fix" even though it really means "Can't Fix." :-/

**Comment by TommyGH** [comment]
I will defer any further comments since brighter programmers than I have already been around the block on this one, but the back of my feeble mind is not comfortable with the phrase "Can't Fix".

**Comment by Michaelangel007** [comment]
I'll put together a demo that shows this problem is _unsolvable_ later ...

**Comment by Michaelangel007** [comment]
The simplest way to show that this problem is unsolvable is:

```
U F93D // top $F93D, cursor on $F961
<Up>   // top $F93A, cursor on $F960
<Up>   // top $F95F, cursor on $F95F
```
 
The cursor will move to the top line because **it is impossible for the debugger to show the cursor line on $F95E in the middle of the screen (zero-based row 21).**

First, let's consider the initial disassembly, top = $F93D: (I've added row numbers)

```
    -4 F936:90 F2    BCC $F92A
    -3 F938:20 56 F9 JSR $F956
    -2 F93B:AA       TAX
    -1 F93C:E8       INX
    -- -----------------------
     0 F93D:D0 01    BNE $F940
     1 F93F:C8       INY
     2 F940:98       TYA
     3 F941:20 DA FD JSR $FDDA
     4 F944:8A       TXA
     5 F945:4C DA FD JMP $FDDA
     6 F948:A2 03    LDX #$03
     7 F94A:A9 A0    LDA #$A0
     8 F94C:20 ED FD JSR $FDED
     9 F94F:CA       DEX
    10 F950:D0 F8    BNE $F94A
    11 F952:60       RTS
    12 F953:38       SEC
    13 F954:A5 2F    LDA $2F
    14 F956:A4 3B    LDY $3B
    15 F958:AA       TAX
    16 F959:10 01    BPL $F95C
    17 F95B:88       DEY
    18 F95C:65 3A    ADC $3A
    19 F95E:90 01    BCC $F961
    20 F960:C8       INY
    21 F961:60       RTS        <-- cursor
```

Press `Up` once, top moved up by 3, top = $F93A

    -2 F936:90 F2    BCC $F92A
    -1 F938:20 56    disasm in middle of JSR!
    -- -----------------------
     0 F93A:F9 AA E8 SBC $E8AA,Y
     1 F93D:D0 01    BNE $F940
     2 F93F:C8       INY
     3 F940:98       TYA
     4 F941:20 DA FD JSR $FDDA
     5 F944:8A       TXA
     6 F945:4C DA FD JMP $FDDA
     7 F948:A2 03    LDX #$03
     8 F94A:A9 A0    LDA #$A0
     9 F94C:20 ED FD JSR $FDED
    10 F94F:CA       DEX
    11 F950:D0 F8    BNE $F94A
    12 F952:60       RTS
    13 F953:38       SEC
    14 F954:A5 2F    LDA $2F
    15 F956:A4 3B    LDY $3B
    16 F958:AA       TAX
    17 F959:10 01    BPL $F95C
    18 F95B:88       DEY
    19 F95C:65 3A    ADC $3A
    20 F95E:90 01    BCC $F961
    21 F960:C8       INY        <-- cursor
    22 F961:60       RTS

So why does the cursor line move to the top of the screen on the next `Up` ?  Because there is NO state where address $F95E is on row 21.

There are _3 possible cases_ to consider when the user presses `Up` when the top line is $F93A.

1. $F93A-1 = $F939

```
    -2 F936:90 F2    BCC $F92A
    -1 F938:20       disasm in middle of JSR
    -- -----------------------
     0 F939:56 F9    LSR $9,X
     1 F93B:AA       TAX
     2 F93C:E8       INX
     3 F93D:D0 01    BNE $F940
     4 F93F:C8       INY
     5 F940:98       TYA
     6 F941:20 DA FD JSR $FDDA
     7 F944:8A       TXA
     8 F945:4C DA FD JMP $FDDA
     9 F948:A2 03    LDX #$03
    10 F94A:A9 A0    LDA #$A0
    11 F94C:20 ED FD JSR $FDED
    12 F94F:CA       DEX
    13 F950:D0 F8    BNE $F94A
    14 F952:60       RTS
    15 F953:38       SEC
    16 F954:A5 2F    LDA $2F
    17 F956:A4 3B    LDY $3B
    18 F958:AA       TAX
    19 F959:10 01    BPL $F95C
    20 F95B:88       DEY
    21 F95C:65 3A    ADC $3A    <-- cursor
    22 F95E:90 01    BCC $F961
    23 F960:C8
```

2. $F93A-2 = $F938

```
    -1 F936:90 F2    BCC $F92A
    -- -----------------------
     0 F938:20 56 F9 JSR $F956
     1 F93B:AA       TAX
     2 F93C:E8       INX
     3 F93D:D0 01    BNE $F940
     4 F93F:C8       INY
     5 F940:98       TYA
     6 F941:20 DA FD JSR $FDDA
     7 F944:8A       TXA
     8 F945:4C DA FD JMP $FDDA
     9 F948:A2 03    LDX #$03
    10 F94A:A9 A0    LDA #$A0
    11 F94C:20 ED FD JSR $FDED
    12 F94F:CA       DEX
    13 F950:D0 F8    BNE $F94A
    14 F952:60       RTS
    15 F953:38       SEC
    16 F954:A5 2F    LDA $2F
    17 F956:A4 3B    LDY $3B
    18 F958:AA       TAX
    19 F959:10 01    BPL $F95C
    20 F95B:88       DEY
    21 F95C:65 3A    ADC $3A    <-- cursor
    22 F95E:90 01    BCC $F961
    23 F960:C8
```
3. $F93A-3 = $F937

```
    -1 F935:90       disasm in middle of BCC
    -- -----------------------
     0 F937:F2 20    SBC ($20)
     1 F939:56 F9    LSR $9,X
     2 F93B:AA       TAX
     3 F93C:E8       INX
     4 F93D:D0 01    BNE $F940
     5 F93F:C8       INY
     6 F940:98       TYA
     7 F941:20 DA FD JSR $FDDA
     8 F944:8A       TXA
     9 F945:4C DA FD JMP $FDDA
    10 F948:A2 03    LDX #$03
    11 F94A:A9 A0    LDA #$A0
    12 F94C:20 ED FD JSR $FDED
    13 F94F:CA       DEX
    14 F950:D0 F8    BNE $F94A
    15 F952:60       RTS
    16 F953:38       SEC
    17 F954:A5 2F    LDA $2F
    18 F956:A4 3B    LDY $3B
    19 F958:AA       TAX
    20 F959:10 01    BPL $F95C
    21 F95B:88       DEY        <-- cursor
    22 F95C:65 3A    ADC $3A
    23 F95E:90 01    BCC $F961
    24 F960:C8
```

The debugger has no way of knowing that `F93A:F9 AA E8 SBC $E8AA,Y` is in the middle of an JSR.

When the top = $F93D, and the cursor was $F961, the debugger could have "guess" that it should only move up by one byte, but that will _still_ cause problems.

**Comment by Michaelangel007** [comment]
> So, is there anything that can be done to help prevent the display from jumping up and down so much? 

Yes, press `Shift+Up` to force the debugger top line to move up by one byte -- this will _help_ keep the debugger in sync:

```
u f93d
<Shift+Up>
<Up>
<Up>
```

**Comment by Michaelangel007** [comment]
As promised here is a demo that shows **this is unsolvable.**

```
0338:20 A9 20
033B:EA      
033C:EA      
033D:EA      
033E:EA      
033F:EA      
0340:EA      
0341:EA      
0342:EA      
0343:EA      
0344:EA      
0345:EA      
0346:EA      
0347:EA      
0348:EA      
0349:EA      
034A:EA      
034B:EA      
034C:EA      
034D:EA      
034E:EA      
034F:A9 50   
0351:A9 52   
0353:EA      
0354:EA      
u 33D
```

1. `Up`
2. `Up`

Could the debugger be made smarter? Yes, but there will be "edge cases" where it fails.
